inventory-api/src/main/scala/com/normation/inventory/services/provisionning/CheckInventoryDigest.scala

```diff
@@ -0,0 +1,170 @@
+import com.normation.inventory.domain._
+
+
+
+/**
+ * We are using a simple date structure that handle the digest file
+ * version and content
+ */
+sealed trait InventoryDigest
+
+final case class InventoryDigestV1(
+    algorithm: String
+  , digest   : String
+) extends InventoryDigest
+
+/**
+ * This trait allow to check digest file for an inventory.
+ * It handles the parsing of the .sig file.
+ * The actual checking is done in CheckInventoryDigest
+ */
+trait ParseInventoryDigestFile {
+  def parse(is: InputStream): Box[InventoryDigest]
+}
+
+/**
+ * Parse a V1 file format:
+ * -------
+ * header=rudder-signature-v1
+ * algorithm=${HASH}
+ * digest=${SIGNATURE}
+ * -------
+ */
+class ParseInventoryDigestFileV1 extends ParseInventoryDigestFile with Loggable {
+  def parse(is: InputStream): Box[InventoryDigest] = {
+
+    val properties = new Properties()
+
+    for {
+      loaded  <- try {
+                   import scala.collection.JavaConverters._
+                   properties.load(is)
+
+                   Full(properties.asInstanceOf[java.util.Map[String, String]].asScala.toMap)
+                 } catch {
+                   case ex: Exception => Failure("Failed to load properties for the signature file", Full(ex), Empty)
+                 }
+      //check version
+
+      v_ok    <- Box((loaded.get("header").filter( _.trim.toLowerCase == "rudder-signature-v1" ))) ?~! "could not read 'header'"
+      algo    <- Box(loaded.get("algorithm").map( _.trim.toLowerCase)) ?~! "could not read 'algorithm'"
+      algo_ok <- if(algo == "sha512") {  // in v1, we only accept sha512
+                   Full("ok")
+                 } else {
+                   Failure(s"The algorithm '${algo}' contains in the digest file is not authorized, only 'sha512' is.")
+                 }
+      digest  <- Box(loaded.get("digest").map( _.trim)) ?~! "could not read 'digest'"
+    } yield {
+      InventoryDigestV1(algo,digest)
+    }
+  }
+}
+
+trait CheckInventoryDigest {
+
+
+  /**
+   * Here, we want to calculate the digest. The good library for that is most likely
+   * bouncy castle: https://www.bouncycastle.org/
+   */
+  def check(pubKey: PublicKey, digest: InventoryDigest, inventoryStream: InputStream): Box[Boolean] = {
+    try {
+      val signature = Signature.getInstance("SHA512withRSA", "BC");
+      signature.initVerify(pubKey);
+      val data = IOUtils.toByteArray(inventoryStream)
+      signature.update(data);
+      digest match {
+        case InventoryDigestV1(_,digest) =>
+        val sig = DatatypeConverter.parseHexBinary(digest)
+
+        Full(signature.verify(sig))
+      }
+    } catch {
+      case e : Exception =>
+        e.printStackTrace()
+        Failure(e.getMessage())
+    }
+  }
+
+}
+
+/**
+ * How we get the key from the inventory received
+ */
+trait GetKey {
+
+  def getKey (receivedInventory  : InventoryReport) : Box[(PublicKey, KeyStatus)]
+
+}
+
+
+
+case class InventoryDigestServiceV1(
```

Comment:
Please, do not use case class for anything else but data structure. The have limit (inheritance related for ex) and provides facilities (copy method) that we don't want on service classes.